pkg/reconcile/pipeline/context/impl.go

```diff
@@ -118,15 +121,43 @@ func (i *impl) Services() ([]pipeline.Service, error) {
 func (i *impl) Applications() ([]pipeline.Application, error) {
 	if i.applications == nil {
 		if ref := i.serviceBinding.Spec.Application; ref != nil {
+			if i.serviceBinding.Spec.Application.Name != "" &&
+				i.serviceBinding.Spec.Application.LabelSelector != nil &&
+				i.serviceBinding.Spec.Application.LabelSelector.MatchLabels != nil {
+				return nil, errs.New("Both label selector and name for application should not be used")
+			}
 			gvr, err := i.typeLookup.ResourceForReferable(ref)
 			if err != nil {
 				return nil, err
 			}
-			u, err := i.client.Resource(*gvr).Namespace(i.serviceBinding.Namespace).Get(context.Background(), ref.Name, metav1.GetOptions{})
-			if err != nil {
-				return nil, err
+			if i.serviceBinding.Spec.Application.Name != "" {
+				u, err := i.client.Resource(*gvr).Namespace(i.serviceBinding.Namespace).Get(context.Background(), ref.Name, metav1.GetOptions{})
+				if err != nil {
+					return nil, err
+				}
+				i.applications = append(i.applications, &application{gvr: gvr, persistedResource: u, bindingPath: i.serviceBinding.Spec.Application.BindingPath})
+			}
+			if i.serviceBinding.Spec.Application.LabelSelector != nil && i.serviceBinding.Spec.Application.LabelSelector.MatchLabels != nil {
+				matchLabels := i.serviceBinding.Spec.Application.LabelSelector.MatchLabels
+				opts := metav1.ListOptions{
+					LabelSelector: labels.Set(matchLabels).String(),
+				}
+
+				emptyResult := make([]pipeline.Application, 0)
+				objList, err := i.client.Resource(*gvr).Namespace(i.serviceBinding.Namespace).List(context.Background(), opts)
+				if err != nil {
+					return emptyResult, err
+				}
+
+				if len(objList.Items) == 0 {
+					return emptyResult, nil
+				}
+
+				for index, _ := range objList.Items {
+					i.applications = append(i.applications, &application{gvr: gvr, persistedResource: &(objList.Items[index]), bindingPath: i.serviceBinding.Spec.Application.BindingPath})
```

Comment:
The `persistedResource` value (address of the item) is taken using the index.
Ref. https://play.golang.org/p/ePs7st6f91g